concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -586,67 +762,89 @@ writeChunks
+    -- |Last finalization record index.
+    FinalizationIndex ->
+    -- |Action for getting the block.
+    GetBlockAt cpv m ->
+    -- |Number of exported blocks and last finalization record index.
+    m (Word64, FinalizationIndex)
+exportBlocksToChunk hdl firstHeight chunkSize lastFinalizationRecordIndex getBlockAt = ebtc firstHeight 0 lastFinalizationRecordIndex
   where
-    ebtc height count =
-        resizeOnResized (readFinalizedBlockAtHeight height) >>= \case
-            Nothing -> return count
-            Just sb | NormalBlock normalBlock <- sbBlock sb -> do
-                let serializedBlock =
-                        runPut $ putVersionedBlock (protocolVersion @pv) normalBlock
-                    len = fromIntegral $ BS.length serializedBlock
-                liftIO $ do
-                    BS.hPut hdl $ runPut $ putWord64be len
-                    BS.hPut hdl serializedBlock
-                forM_ (blockFields (sbBlock sb)) $ \bf ->
-                    case blockFinalizationData bf of
-                        BlockFinalizationData fr ->
-                            dbsLastFinIndex .= finalizationIndex fr
-                        NoFinalizationData -> return ()
-                if count < chunkSize
-                    then ebtc (height + 1) (count + 1)
-                    else return count
-            Just _ -> return count -- this branch should never be reachable, genesis blocks are not exported.
+    ebtc :: BlockHeight -> Word64 -> FinalizationIndex -> m (Word64, FinalizationIndex)
+    ebtc height count lastFinRecIdx = case getBlockAt of
+        GetBlockAtV0 f ->
+            f height >>= \case
+                Nothing -> return (count, lastFinRecIdx)
+                Just (serializedBlock, finalizationRecordIndexM) -> do
+                    void $ writeBlockOut serializedBlock
+                    -- In case there wasn't a @FinalizationRecord@ present in the block
+                    -- we carry over the latest seen @FinalizationIndex@.
+                    continue count height $ fromMaybe lastFinRecIdx finalizationRecordIndexM
+        GetBlockAtV1 f ->
+            f height >>= \case
+                Nothing -> return (count, 0)
+                Just serializedBlock -> do
+                    void $ writeBlockOut serializedBlock
+                    -- We simply set the last @FinalizationRecord@ index to 0,
+                    -- as they are not a concept for 'ConsensusV1'.
+                    continue count height 0
+    writeBlockOut :: BS.ByteString -> m ()
+    writeBlockOut serializedBlock = do
+        let len = fromIntegral $ BS.length serializedBlock
+        liftIO $ do
+            BS.hPut hdl $ runPut $ putWord64be len
+            BS.hPut hdl serializedBlock
+    continue count height finalizationRecordIndex =
+        if count < chunkSize
+            then ebtc (height + 1) (count + 1) finalizationRecordIndex
+            else return (count, finalizationRecordIndex)
```

Comment:
I would suggest making separate V0 and V1 versions of this function, rather than combining them together as currently done. In particular, this avoids casing on `getBlockAt` inside the loop, and we can do without finalization indexes altogether in V1 (where they don't exist as a concept). Of course, the likes of `writeBlockOut` can be factored out.

Reply:
Same as above.